packages/framework/src/Console/Helpers/InteractivePublishCommandHelper.php

```diff
@@ -0,0 +1,143 @@
+ * @internal This class offloads logic from the PublishViewsCommand class and should not be used elsewhere.
+ */
+class InteractivePublishCommandHelper
+{
+    protected readonly string $group;
+    protected readonly string $sourceDirectory;
+    protected readonly string $targetDirectory;
+
+    /** @var array<string, string> Map of source files to target files */
+    protected readonly array $publishableFilesMap;
+
+    /** @param "hyde-layouts"|"hyde-components"|"hyde-page-404" $group */
+    public function __construct(string $group)
+    {
+        $this->group = $group;
+
+        [$this->sourceDirectory, $this->targetDirectory] = $this->getPublishPaths();
+
+        $this->publishableFilesMap = $this->mapPublishableFiles($this->findAllFilesForTag());
+    }
+
+    /** @return array<string, string> */
+    public function getFileChoices(): array
+    {
+        return Arr::mapWithKeys($this->publishableFilesMap, /** @return array<string, string> */ function (string $source): array {
+            return [$source => $this->pathRelativeToDirectory($source, $this->targetDirectory)];
+        });
+    }
+
+    /** @param array<string> $selectedFiles */
+    public function handle(array $selectedFiles): void
+    {
+        $filesToPublish = $this->filterPublishableFiles($selectedFiles);
+
+        $this->publishFiles($filesToPublish);
+    }
+
+    /** @return array{string, string} */
+    protected function getPublishPaths(): array
+    {
+        $viewPaths = ServiceProvider::pathsToPublish(ViewServiceProvider::class, $this->group);
+
+        $source = array_key_first($viewPaths);
+        $target = $viewPaths[$source];
+
+        return [$source, $target];
+    }
+
+    /** @return \Symfony\Component\Finder\SplFileInfo[] */
+    protected function findAllFilesForTag(): array
+    {
+        return File::allFiles($this->sourceDirectory);
+    }
+
+    /**
+     * @param  \Symfony\Component\Finder\SplFileInfo[]  $search
+     * @return array<string, string>
+     */
+    protected function mapPublishableFiles(array $search): array
+    {
+        return Arr::mapWithKeys($search, /** @return array<string, string> */ function (SplFileInfo $file): array {
+            $targetPath = path_join($this->targetDirectory, $file->getRelativePathname());
+
+            return [Hyde::pathToRelative(realpath($file->getPathname())) => Hyde::pathToRelative($targetPath)];
+        });
+    }
+
+    /** @param array<string, string> $selectedFiles */
+    protected function publishFiles(array $selectedFiles): void
+    {
+        foreach ($selectedFiles as $source => $target) {
+            Filesystem::ensureDirectoryExists(dirname($target));
+            Filesystem::copy($source, $target); // Todo: See how we should handle existing files
+        }
+    }
+
+    /**
+     * @experimental This method may be toned down in the future.
+     *
+     * @param  array<string>  $selectedFiles
+     */
+    public function formatOutput(array $selectedFiles): string
```

Comment:
Is this too far?


All files (10): "Published all files, including [app.blade.php, docs.blade.php, footer.blade.php] and 7 more"
Many files (7): "Published files [app.blade.php, docs.blade.php, footer.blade.php] and 4 more"
Multiple files (3): "Published files [app.blade.php, docs.blade.php, footer.blade.php]"
Single file (1): "Published file [app.blade.php]"